src/duplex.rs

```diff
@@ -78,48 +79,90 @@ pub struct DuplexStream {
 
 impl Drop for DuplexStream {
     fn drop(&mut self) {
-        futures::executor::block_on(async {
-            // notify the other side of the closure
-            self.write.lock().await.close_write();
-            self.read.lock().await.close_read();
-        })
+        // notify the other side of the closure
+        self.write.lock().unwrap().close_write();
+        self.read.lock().unwrap().close_read();
     }
 }
 
+/// A helper macro to `.await` the I/O function, used in our I/O traits
+/// implementations.
+///
+/// Different from a plain `function().await`, it drops the lock guard if future
+/// `function()` is pending.
+macro_rules! await_io_future {
+    // For functions that do not have arugments: flush/shutdown
+    ($trait:ident, $function:ident, $guard:expr) => {{
+        let opt_read_ready = <SimplexStream as $trait>::$function(&mut *$guard).now_or_never();
+
+        match opt_read_ready {
+            Some(result) => result,
+            None => {
+                // drop the Mutex guard or it could deadlock
+                // https://github.com/SteveLauC/monoio-duplex/issues/7
+                drop($guard);
+                std::future::pending().await
```

Comment:
Future polls to this future will always return Pending due to this line, the state of this future has been changed, which should never be done!